projecttacoma/fqm-execution
# Or gap grouping

# Summary
This Pull Requests adds (very basic) support for "or gaps" in the gaps-in-care calculator. 

An Or gap is when any of several gaps in care would, by itself, put the patient into the desired population (usually the numerator) if addressed. The converse is an And gap (which we define as any non-Or gap, at the moment), where you need each of several gaps to be addressed to put the patient into the desired population.

At the moment, our Or gap checking is based on the `ExpressionStack` on each `DataTypeQuery`. We consider gaps to be "or-d" if they share a top-level `ExpressionStackEntry` (aka same library and local ID) and that shared Entry is of type `Or`.

We also re-arranged the Gaps output to match our proposed Grouped Gaps structure, seen in the image below: 
<img width="1394" alt="Screen Shot 2021-04-21 at 11 45 10 AM" src="https://user-images.githubusercontent.com/49001/115582557-15891980-a297-11eb-9dd4-dccaff9f728a.png">

"or" gaps are represented by a single `DetectedIssue` resource, with multiple `GuidanceResponse`s, indicating that any of the nested `GuidanceResponse`s would satisfy the `DetectedIssue`. "And" gaps are represented by multiple `DetectedIssue` resources, each with a single `GuidanceResponse`. 

This PR will, in theory, support multiple-"or" gaps and "and" gaps at the top level. If there are (for example) two sets of "or-d" gaps, each with two ways to satisfy the Or, you would end up with two `DetectedIssue` resources, each with two `GuidanceResponse` resources nested below them.

We do not yet support nested "and" and "or" gaps.

## New behavior
* Function added to "Group" gaps as "or" gaps where appropriate. Currently very basic, could get more complex in the future.
* Generation of the `ExpressionStack` element on `DataTypeQuery` fixed to not truncate stack entries from parent queries.
* Output of gaps-in-care calculation updated to properly follow the `Representing Grouped Gaps` diagram above. NOTE: This is a breaking change for most measures not named EXM130.

## Code changes
* Added `groupGapQueries` function, which generates an array of arrays of `DataTpeQuery` objects, which will then get turned into the new gap structure
* Updated gap tests to follow the new gaps output structure
* Added `isNearMiss` to the `DataTypeQuery` interface definition.

# Testing guidance
* Run EXM130 using a denominator patient (the "test-denom" patient out of `connectathon` works well) and the measurement period start and end dates of `2019-01-01` and `2019-12-31`. 
* Examine the output Bundle. There should be only one `DetectedIssue` contained in it (because all the missing elements in EXM130 are "or'd"), with several `GuidanceResposne` elements nested as `evidence.detail`
* Run some other measure that doesn't necessarily have OR'd gaps (EXM104 for example). You should see multiple `DetectedIssue` resources in the output `Bundle`.
